### PR TITLE
PAS-115B courseItem API calls bugfix

### PR DESCRIPTION
**Pull request by isaiahdoyle**
This is the backend portion of the small changes I had to make in order to get the front-end API calls to work while testing.

### CODE REVIEW

File: rollup.config.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import typescript from '@rollup/plugin-typescript';
 import { nodeResolve } from '@rollup/plugin-node-resolve';
-import commonjs from '@rollup/plugin-commonjs';
+// import commonjs from '@rollup/plugin-commonjs';
```
Comment (BraidenCutforth) [inline rollup.config.js:3]: You can just remove this

File: rollup.config.js [hunk 2]
```diff
@@ -9,5 +9,6 @@ export default {
 		file: 'dist/index.js',
 		format: 'cjs',
 	},
-	plugins: [json(), typescript(), commonjs(), nodeResolve()],
+	// plugins: [json(), typescript(), commonjs(), nodeResolve()],
```
Comment (BraidenCutforth) [inline rollup.config.js:12]: You can remove the commented out code.

File: src/models/courseItem.ts [hunk 3]
```diff
@@ -11,7 +11,7 @@ export const courseItemSchema = new dynamoose.Schema(
 		name: String,
 		weight: Number,
 		grade: [Number, fraction],
-		dueDate: Date,
+		date: [String, Date],
```
Comment (BraidenCutforth) [inline src/models/courseItem.ts:14]: Just wondering why this change was necessary. If when we load the data from the database it is returned as a string initially we can modify it to be a date immediately if needed, though it will end being turned into a string for the request I think.

File: src/operators/courseItemOperations.ts [hunk 4]
```diff
@@ -7,8 +7,21 @@ const checkCourseItemUser = checkUserId(CourseItem.get);
 export const create = async (courseItem: CourseItemInterface) => {
 	const hashKey = uuidv4();
 	courseItem.id = hashKey;
-	// logs uuid of new courseItem for testing
-	console.log(courseItem.name + ' id: ' + courseItem.id);
+	// logs uuid of new courseItem [[ for testing ]]
+	// console.log(courseItem.name + ' id: ' + courseItem.id);
+	// console.log(courseItem.name + ' grade: ' + courseItem.grade);
```
Comment (BraidenCutforth) [inline src/operators/courseItemOperations.ts:12]: You can remove unless you feel there is value in keeping it for testing.

File: src/routers/courseItem.ts [hunk 5]
```diff
@@ -22,13 +22,26 @@ cItemRouter.get('/:id', async (req, res) => {
 	}
 });
 
+// helper function converts string to number
+//   - should be updated to support fractions
+const numberify = (str: string) => {
+	if (str) {
+		return Number(str);
+	}
+};
```
Comment (BraidenCutforth) [inline src/routers/courseItem.ts:31]: If you want to support `0.66` and support `10/15` you can do this:

```suggestion
const numberify = (str: string) => {
     const regex = /^(\d+\.?\d*)\/(\d+\.?\d*)$/;
     const match = str.match(regex);
     if(match) {
         return Number(match[1])/Number(match[2]);
     }
     
     return Number(str); // Not guranteed to be a number since we don't know what is store in str
};
```

File: src/routers/courseItem.ts [hunk 6]
```diff
@@ -22,13 +22,26 @@ cItemRouter.get('/:id', async (req, res) => {
...
+
 cItemRouter.post('/', async (req, res) => {
 	try {
 		const userID = req.header('userID');
 		if (userID) {
-			await update(req.body, userID);
+			const body = req.body;
+			body.owner = userID;
+			body.weight = numberify(body.weight);
+			body.grade = numberify(body.grade);
```
Comment (BraidenCutforth) [inline src/routers/courseItem.ts:40]: So I realize we can make this a bit better by following some of the auggestions in this [SO answer](https://stackoverflow.com/questions/25983090/is-sanitizing-json-necessary). Specifically, do 

```js
const courseItem = {
    owner: userId;
    weight: numberify(body.weight);
    // other wanted properties ...
}
```

This helps us avoid adding properties that someone maliciously adds to the request etc.

File: src/routers/courseItem.ts [hunk 7]
```diff
@@ -37,9 +50,12 @@ cItemRouter.post('/', async (req, res) => {
 
 cItemRouter.put('/', async (req, res) => {
 	try {
-		await create(req.body);
-		console.log('Put Course Item');
-		res.send('Put cItemRouter: ' + req.body.name);
+		const userID = req.header('userID');
+		if (userID) {
+			await update(req.body, userID);
```
Comment (BraidenCutforth) [inline src/routers/courseItem.ts:55]: We should probably build the CourseItem from the req.body in the same way as above (extract to method 😄)

**Review by BraidenCutforth — CHANGES_REQUESTED** [review]

File: src/models/courseItem.ts (hunk 3 above)
Comment (isaiahdoyle) [reply src/models/courseItem.ts:14]: Honestly I just changed this in order to make sure the calls worked in the first place - this should definitely be changed back, it was just a way to deal with typing issues without actually solving the problem in the first place (in other words, I was too lazy to actually convert the date from the add item modal into a date). Should I try and solve the issue on this ticket? Or would it be best in the scope of a smaller ticket?

File: src/routers/courseItem.ts [hunk 8]
```diff
@@ -22,13 +22,45 @@ cItemRouter.get('/:id', async (req, res) => {
...
+				conformToSchema: body.conformToSchema,
+				toDynamo: body.toDynamo,
+				prepareForResponse: body.prepareForResponse,
+				original: body.original,
+				toJSON: body.toJSON,
+				serialize: body.serialize,
+				delete: body.delete,
+				save: body.save,
+				populate: body.populate,
+			};
```
Comment (isaiahdoyle) [inline src/routers/courseItem.ts:60]: @BraidenCutforth is this close to what you had in mind for sanitizing the input JSON? In order for `create` to work it needs ALL the data from `courseItemInterface`. Thoughts?

**Review by cindyunrau — APPROVED** [review]

File: src/routers/courseItem.ts (hunk 8 above)
Comment (BraidenCutforth) [reply src/routers/courseItem.ts:60]: Per conversation we probably need to call this function to create the document so we can remove all keys except the actual data ones.

```suggestion
			const courseItem = new CourseItemModel({
				id: body.id,
				owner: userID,
				name: body.name,
				weight: numberify(body.weight),
				grade: numberify(body.grade),
				date: body.date,
				createdAt: body.createdAt,
				updatedAt: body.updatedAt,
				conformToSchema: body.conformToSchema,
				toDynamo: body.toDynamo,
				prepareForResponse: body.prepareForResponse,
				original: body.original,
				toJSON: body.toJSON,
				serialize: body.serialize,
				delete: body.delete,
				save: body.save,
				populate: body.populate,
			});
```

**Review by BraidenCutforth — CHANGES_REQUESTED** [review]

File: src/routers/courseItem.ts (hunk 8 above)
Comment (isaiahdoyle) [reply src/routers/courseItem.ts:60]:
```suggestion
			const courseItem = {
				id: body.id,
				owner: userID,
				name: body.name,
				weight: numberify(body.weight),
				grade: numberify(body.grade),
				date: body.date,
				createdAt: body.createdAt,
				updatedAt: body.updatedAt,
			};
```
Did you mean this?

Comment (BraidenCutforth) [reply src/routers/courseItem.ts:60]: Yup! Sorry about that, forgot to remove the unneeded fields!

File: src/operators/courseItemOperations.ts [hunk 9]
```diff
@@ -7,8 +7,18 @@ const checkCourseItemUser = checkUserId(CourseItem.get);
...
+	// if grade, weight, and/or date fields are empty, set to undefined
+	if (!courseItem.grade) {
+		courseItem.grade = undefined;
+	}
+	if (!courseItem.weight) {
+		courseItem.weight = undefined;
+	}
+	if (!courseItem.date) {
+		courseItem.date = undefined;
+	}
```
Comment (BraidenCutforth) [inline src/operators/courseItemOperations.ts:20]: Check on this and see if this is the behaviour we want, I know grade/weight being `0` would cause issues.

**Review by BraidenCutforth — APPROVED** [review]
Just check over what issues we are actually trying to fix in the data for the create function.
